### PR TITLE
feat: Add typed variation methods.

### LINKED ISSUE #285: Typesafe API for TypeScript consumers
**Is your feature request related to a problem? Please describe.**

An API for TypeScript users to retrieve toggle values in a typesafe manner.

**Describe the solution you'd like**

Similar to the Java SDK, provide methods like `boolVariation` and `stringVariation` on `LDClient` for TypeScript users that only accept a default value of the correct type, and throw an error if the resolved value from LaunchDarkly is not of the correct type.

### PR DESCRIPTION
**Pull request by kinyoklion**
This adds typed variation methods. 

This works for both JS and TS, but may be especially nice for those that want to be more type safe in typescript.

resolves #285

### CODE REVIEW

File: contract-tests/index.js [hunk 1]
```diff
@@ -31,7 +31,8 @@ app.get('/', (req, res) => {
       'migrations',
       'event-sampling',
       'config-override-kind',
-      'metric-kind'
+      'metric-kind',
+      'strongly-typed',
```
Comment (kinyoklion) [inline contract-tests/index.js:35]: Run the typed contract tests.

File: contract-tests/sdkClientEntity.js [hunk 2]
```diff
@@ -124,10 +124,32 @@ export async function newSdkClientEntity(options) {
       case 'evaluate': {
         const pe = params.evaluate;
         if (pe.detail) {
-          return await client.variationDetail(pe.flagKey, pe.context || pe.user, pe.defaultValue);
+          switch(pe.valueType) {
+            case "bool":
+              return await client.boolVariationDetail(pe.flagKey, pe.context || pe.user, pe.defaultValue);
+            case "int":
```
Comment (kinyoklion) [inline contract-tests/sdkClientEntity.js:130]: The contract tests support int and double, but we are using a single numberVariation.

File: packages/shared/sdk-server/__tests__/LDClient.evaluation.test.ts [hunk 3]
```diff
@@ -160,6 +160,99 @@ describe('given an LDClient with test data', () => {
     const valueB = await client.variation('my-feature-flag-1', userContextObject, 'default');
     expect(valueB).toEqual(true);
   });
+
+  it('evaluates with jsonVariation', async () => {
```
Comment (kinyoklion) [inline packages/shared/sdk-server/__tests__/LDClient.evaluation.test.ts:164]: This just addresses the problem with `any` in Typescript. Some people require using `unknown` so they have to explicitly cast.

File: packages/shared/sdk-server/__tests__/LDClient.evaluation.test.ts [hunk 4]
```diff
@@ -160,6 +160,99 @@ describe('given an LDClient with test data', () => {
...
+      defaultUser,
+      false,
+    )) as number;
+    expect(numericRes).toBe(62);
+
+    td.update(td.flag('flagkey').valueForAll('potato'));
+    const stringRes: string = (await client.jsonVariation('flagkey', defaultUser, false)) as string;
+    expect(stringRes).toBe('potato');
+  });
+
```
Comment (kinyoklion) [inline packages/shared/sdk-server/__tests__/LDClient.evaluation.test.ts:181]: Below are all the typed method calls.

File: packages/shared/sdk-server/__tests__/LDClient.migrations.test.ts [hunk 5]
```diff
@@ -57,7 +57,7 @@ describe('given an LDClient with test data', () => {
       const defaultValue = Object.values(LDMigrationStage).find((item) => item !== value);
       // Verify the pre-condition that the default value is not the value under test.
       expect(defaultValue).not.toEqual(value);
-      const res = await client.variationMigration(
+      const res = await client.migrationVariation(
```
Comment (kinyoklion) [inline packages/shared/sdk-server/__tests__/LDClient.migrations.test.ts:60]: I decided to make this the same as the other SDKs with this change. This has not been released yet (for Yus).

File: packages/shared/sdk-server/src/LDClientImpl.ts [hunk 6]
```diff
@@ -301,7 +303,105 @@ export default class LDClientImpl implements LDClient {
...
+          };
+          resolve(typedRes);
+        },
+        typeChecker,
+      );
+    });
+  }
+
+  async boolVariation(key: string, context: LDContext, defaultValue: boolean): Promise<boolean> {
+    return (
```
Comment (kinyoklion) [inline packages/shared/sdk-server/src/LDClientImpl.ts:333]: The approach taken with typedEval is to have it take a validator that checks the type and also indicates what the type was, for error messages. This matched nicely with the TypeValidators implementation. 

If the type is wrong, then it generates the correct reason and evaluation events.

File: contract-tests/sdkClientEntity.js [hunk 7]
```diff
@@ -124,10 +124,32 @@ export async function newSdkClientEntity(options) {
...
+              return await client.numberVariationDetail(pe.flagKey, pe.context || pe.user, pe.defaultValue);
+            case "double":
+              return await client.numberVariationDetail(pe.flagKey, pe.context || pe.user, pe.defaultValue);
```
Comment (yusinto) [inline contract-tests/sdkClientEntity.js:133]:
```suggestion
            case "int":
            case "double":
              return await client.numberVariationDetail(pe.flagKey, pe.context || pe.user, pe.defaultValue);
```

File: contract-tests/sdkClientEntity.js [hunk 8]
```diff
@@ -124,10 +124,32 @@ export async function newSdkClientEntity(options) {
...
         } else {
-          const value = await client.variation(pe.flagKey, pe.context || pe.user, pe.defaultValue);
-          return { value };
+          switch(pe.valueType) {
+            case "bool":
+              return {value: await client.boolVariation(pe.flagKey, pe.context || pe.user, pe.defaultValue)};
+            case "int":
+              return {value: await client.numberVariation(pe.flagKey, pe.context || pe.user, pe.defaultValue)};
+            case "double":
+              return {value: await client.numberVariation(pe.flagKey, pe.context || pe.user, pe.defaultValue)};
```
Comment (yusinto) [inline contract-tests/sdkClientEntity.js:147]: You can waterfall these too like above.

**Review by yusinto — APPROVED** [review]

**Pull request merged**